core/templates/dev/head/services/StateTopAnswersStatsService.js

```diff
@@ -42,32 +54,102 @@ oppia.factory('StateTopAnswersStatsService', [
       var interactionRulesService = $injector.get(
         AngularNameService.getNameOfInteractionRulesService(
           state.interaction.id));
-      stateTopAnswerStatsCache[stateName].forEach(function(answerStats) {
+      var allAnswersCacheEntry =
+        stateTopAnswersStatsCache[stateName].allAnswers;
+      var unresolvedAnswersCacheEntry =
+        stateTopAnswersStatsCache[stateName].unresolvedAnswers;
+
+      // Clear the unresolved answers array since many answers may now have
+      // different values.
+      unresolvedAnswersCacheEntry.splice(0, unresolvedAnswersCacheEntry.length);
+
+      // Update the isAddressed data of each answer and put any unaddressed
+      // answers into the unresolvedAnswers array.
+      allAnswersCacheEntry.forEach(function(answerStats) {
         answerStats.isAddressed =
           AnswerClassificationService.isClassifiedExplicitlyOrGoesToNewState(
             explorationId, stateName, state, answerStats.answer,
             interactionRulesService);
+
+        if (!answerStats.isAddressed) {
+          unresolvedAnswersCacheEntry.push(answerStats);
+        }
       });
     };
 
+    var onStateAdded = function(stateName) {
+      stateTopAnswersStatsCache[stateName] = {
+        allAnswers: [],
+        unresolvedAnswers: []
+      };
+    };
+
+    var onStateDeleted = function(stateName) {
+      delete stateTopAnswersStatsCache[stateName];
+    };
+
+    var onStateRenamed = function(oldStateName, newStateName) {
+      stateTopAnswersStatsCache[newStateName] =
+        stateTopAnswersStatsCache[oldStateName];
+      delete stateTopAnswersStatsCache[oldStateName];
+    };
+
+    var onStateInteractionAnswerGroupsSaved = function(stateName) {
+      refreshAddressedInfo(stateName);
+    };
+
+    $rootScope.$on(STATE_ADDED_EVENT_NAME, function(event, args) {
```

Comment:
$scope.$on, ditto below

($rootScope events get propagated downwards)

Also I think we use "evt" instead of "event", the latter is a keyword in IE IIRC